### PR TITLE
[PROF-2711] Fix multiple profilers starting in the same process

### PR DESCRIPTION
**Pull request by ivoanjo**
While experimenting with Ruby profiler, I noticed that sometimes more
than one profiler was being started.

Upon closer investigation, I identified this as a race between the
profiler's sampling thread and the thread that was initializing the
Datadog APM: If initialization of components was not "fast enough",
the Datadog::Profiling::Collectors::Stack would run and try to access
`Datadog.tracer` (stack.rb:140), which would trigger a new
initialization of the components, including starting a second profiler.

This can be triggered reliably by injecting an artificial slow down
into component initialization -- change configuration.rb from

```ruby
    def components
      @components ||= build_components(configuration)
    end
```

to

```ruby
    def components
      @components ||= build_components(configuration).tap { sleep 0.1 }
    end
```

and then by running a trivial application:

```ruby
$ DD_PROFILING_ENABLED=true bundle exec bin/ddtracerb exec ruby -e 'sleep 1; puts Thread.list.count'
203
$ DD_PROFILING_ENABLED=true bundle exec bin/ddtracerb exec ruby -e 'sleep 1; puts Thread.list.count'
195
```

Every initialization starts a new profiler, which starts two new
threads, and due to the race, many more threads get started than
expected -- and the value is not deterministic.

To fix this, I've added a mutex to protect all access to the
`@components`. Thus, adding the same `.tap { sleep 0.1 }` now
yields the three expected threads (one user code, two for the
profiler):

```ruby
$ DD_PROFILING_ENABLED=true bundle exec bin/ddtracerb exec ruby -e 'sleep 1; puts Thread.list.count'
3
```

### CODE REVIEW

File: lib/ddtrace/configuration.rb [hunk 1]
```diff
@@ -9,6 +10,18 @@ module Datadog
   module Configuration
     extend Forwardable
 
+    # Used to ensure that @components initialization/reconfiguration is performed one-at-a-time, by a single thread.
+    #
+    # This is important because components can end up being accessed from multiple application threads (for instance on
+    # a threaded webserver), and we don't want their initialization to clash (for instance, starting two profilers...).
+    #
+    # Note that a Mutex **IS NOT** reentrant: the same thread cannot grab the same Mutex more than once.
+    # This means below we are careful not to nest calls to methods that grab the lock.
```
Comment (marcotc) [inline lib/ddtrace/configuration.rb:19]: Currently our trace does perform reentrant calls to component initialization.
One example is calling `Datadog.logger` when any part of the component initialization reports an error message.

If we want to introduce this change today, I don't think we can have this mutex as non-reentrant: the it is a requirement for `ddtrace` to not prevent the host application from continuing its work, even when `ddtrace` fails.

**Review by marcotc — COMMENTED** [review]
This is a recurring issue with `ddtrace` so it's great that we are addressing it now. One recent example of an user encountering this error is https://github.com/DataDog/dd-trace-rb/issues/1306.

I'd like to see at least one test that fails today because of a realistic concurrent initialization pattern: this could be either the use case that triggered you to create this PR or the same case that #1306 encountered. Here's how to reproduce https://github.com/DataDog/dd-trace-rb/issues/1306#issuecomment-767649779, if relevant.

I left a comment regarding making this safe, even on tracer failure. Let me know if you want to discuss this further.

**Comment by marcotc** [comment]
BTW, the mysql test failures in CI has been fixed in master by https://github.com/DataDog/dd-trace-rb/commit/b2353cfeba163d1f021d9290c277361a8b9c2e82.

File: lib/ddtrace/configuration.rb (hunk 1 above)
Comment (ivoanjo) [reply lib/ddtrace/configuration.rb:19]: Yeah, I'll admit I hesitated a bit between going for `Mutex` or `Monitor`.

I believe the `logger` example you quoted cannot happen -- the `logger`  is built to never trigger the initialization of the components (`if current_components` will never be `true` until the components are properly and fully initialized).

> If we want to introduce this change today, I don't think we can have this mutex as non-reentrant: the it is a requirement for ddtrace to not prevent the host application from continuing its work, even when ddtrace fails.

If the use-case is "it should never break customer code", I was considering an alternative -- what if we caught the resulting `ThreadError` and just logged an error message instead?

I'm somewhat concerned that going with the [thread-safe `Monitor`](https://rubyapi.org/3.0/o/monitor) instead may just silently hide issues, whereas if we handled the exception and printed an error we could identify issues a lot easier.

**Comment by ivoanjo** [comment]
Thank you! I'll work on a rebase of the profiling branch to include your fix, the random failures are hella annoying :)

File: lib/ddtrace/configuration.rb (hunk 1 above)
Comment (marcotc) [reply lib/ddtrace/configuration.rb:19]: We can try catching `ThreadError`, that might do it. 👍

Comment (ivoanjo) [reply lib/ddtrace/configuration.rb:19]: Added handling of `ThreadError` in fdc0d5848b6af87c754161b61d7ad79795e70fbe, let me know your thoughts!

Comment (ivoanjo) [reply lib/ddtrace/configuration.rb:19]: In particular, this is the second time (the other being https://github.com/DataDog/dd-trace-rb/pull/1310) that I've added a user-visible "dear user, oops we messed up! do you mind telling us about it so we can fix it?" message.

I think being clear/open about these allows us to better spot and improve the codebase, but this is a somewhat new pattern in dd-trace-rb so I'm curious what the rest of the team thinks about it.

File: lib/ddtrace/configuration.rb [hunk 2]
```diff
@@ -104,13 +98,28 @@ def reset!
...
     end
 
     private
 
+    def safely_synchronize
+      COMPONENTS_LOCK.synchronize do
+        begin
+          yield
+        rescue ThreadError => e
+          logger_without_components.warn(
```
Comment (marcotc) [inline lib/ddtrace/configuration.rb:113]: I think this is critical enough to be considered an error, given the tracer will be partially initialized when this happens.

Comment (ivoanjo) [reply lib/ddtrace/configuration.rb]: Updated in f6626c9608627cb6f1b04c8bf804426996830b4f

**Comment by ivoanjo** [comment]
Note: I needed to do a force push after `feature/profiling` got rebased. Long-lived branches :((((((

**Review by delner — COMMENTED** [review]
@ivoanjo and I had a chat about this one; overall the strategy seems fine to me!

Just had some questions about performance of `safely_synchronize` and how it would impact the average case of each span in every thread calling `Datadog.tracer` thereby invoking `Configuration#components` and requesting the lock. Would this have significant cost at scale?

If no, then this seems good! If yes, is there a good way to mitigate that cost for the average case? Preventing the race condition is the most important priority here, however if there's a way of keeping overhead to a minimum without adding too much complexity/risk then that would be of good value as well.

**Comment by ivoanjo** [comment]
Note: I needed to do a second force push after rebasing `feature/profiling`.

**Comment by ivoanjo** [comment]
Ok, so here's my findings around the performance impact of this. I've written two benchmarks, which I share below. 

Benchmarks were taken on my laptop (i7-1068NG7, quad-core) using native Ruby for macOS. I've tested with a few Ruby versions to give us coverage and to validate that the results didn't change a lot between versions.

TL;DR there is a measurable impact in a "worst-case" benchmark, and I suggest an alternative below, but let me know what you think.

#### Single-threaded benchmark

```ruby
require 'benchmark/ips'
require 'ddtrace'

raise 'No tracer component available' unless Datadog.tracer.enabled

Benchmark.ips do |x|
  x.time = 10
  x.warmup = 2

  x.report("#{RUBY_VERSION} configuration #{(Datadog::Configuration.const_get(:COMPONENTS_LOCK) rescue false)? "with mutex" : "without mutex"}") do
    Datadog.tracer
  end

  # Compare the iterations per second of the various reports
  x.save! ENV['SAVE_FILE'] if ENV['SAVE_FILE']
  x.compare!
end
```

Results:
```
2.3.8 configuration without mutex:  3917051.6 i/s
3.0.0 configuration without mutex:  3766124.8 i/s - 1.04x  (± 0.00) slower
2.0.0 configuration without mutex:  3166432.5 i/s - 1.24x  (± 0.00) slower
2.3.8 configuration with mutex:  2730354.2 i/s - 1.43x  (± 0.00) slower
3.0.0 configuration with mutex:  2665619.1 i/s - 1.47x  (± 0.00) slower
2.0.0 configuration with mutex:  2274639.5 i/s - 1.72x  (± 0.00) slower
```

**Notes**: With a single thread, being able to grab the tracer more than 2 million times in the worst case (2.0.0 with mutex) seems reasonable-ish? How many times do we expect a reasonable application to grab the tracer per second?

#### Multi-threaded benchmark

```ruby
require 'benchmark/ips'
require 'ddtrace'

raise 'No tracer component available' unless Datadog.tracer.enabled

Benchmark.ips do |x|
  x.time = 10
  x.warmup = 2

  x.report("#{RUBY_VERSION} multithreaded configuration #{(Datadog::Configuration.const_get(:COMPONENTS_LOCK) rescue false)? "with mutex" : "without mutex"}") do
    (1..100).map {
      Thread.new do
        2500.times do
          Datadog.tracer
        end
      end
    }.map(&:join)
  end

  # Compare the iterations per second of the various reports
  x.save! ENV['SAVE_FILE'] if ENV['SAVE_FILE']
  x.compare!
end
```

Results:
```
2.3.8 multithreaded configuration without mutex:       14.8 i/s
3.0.0 multithreaded configuration without mutex:       13.9 i/s - same-ish: difference falls within error
3.0.0 multithreaded configuration with mutex:       10.1 i/s - 1.47x  (± 0.00) slower
2.0.0 multithreaded configuration without mutex:        9.7 i/s - 1.53x  (± 0.00) slower
2.3.8 multithreaded configuration with mutex:        5.5 i/s - 2.69x  (± 0.00) slower
2.0.0 multithreaded configuration with mutex:        3.7 i/s - 4.02x  (± 0.00) slower
```

**Notes**: Hey, we found our Ruby 3x3! 3.0.0 with mutex is 3 times faster than 2.0.0 with mutex!
Kidding aside, there's definitely an impact, but this is quite a big corner case -- 100 threads, each doing nothing but looking and trying to get the `tracer`.

#### Suggestion

One possible alternative would be only use the lock when mutating/causing side effects on the components (creating, shutting down), and skip the lock when just reading the components.

The problem with skipping the lock for reading is that Ruby does not define semantics for [safe initialization nor safe publication](https://shipilev.net/blog/2014/safe-public-construction/) -- e.g. if thread 1 writes to `@components` inside the lock, and thread 2 observes `@components` without grabbing the lock, will the read of thread 2 be correct (e.g. it will see the components correctly initialized).

Under MRI, the read of thread 2 is correct. According to the TruffleRuby developers, under JRuby the read of thread 2 can fail, and in TruffleRuby it works, but it's "kind of an implementation detail".

(from the graalVM slack)
Screen Shot 2021-02-10 at 14 28 14

Thus, what we could do is skip the thread, but only for MRI:

```ruby
    def components
      if RUBY_ENGINE == "ruby"
        existing_components = @components
        return existing_components if existing_components
      end

      safely_synchronize do
        @components ||= build_components(configuration)
      end
    end
```

It's a bit dirty, but it would give us back some of the performance. We'd still be open to weird issues if people try to use any components while they're being shut down / reconfigured, but that's no worse than what we have today, and well, it is a quite corner case. (And of course I'd write a 10 line comment next to the code explaining what that was and why people should be very careful before trying to "optimize" further.

#### Appendix

To test without the mutex, I just commented out the mutex creation and acquisition:

```diff
diff --git a/lib/ddtrace/configuration.rb b/lib/ddtrace/configuration.rb
index f3ff1b33..4cfd2f47 100644
--- a/lib/ddtrace/configuration.rb
+++ b/lib/ddtrace/configuration.rb
@@ -20,8 +20,8 @@ module Datadog
     #
     # Every method that directly or indirectly accesses/mutates @components should be holding the lock (through
     # #safely_synchronize) while doing so.
-    COMPONENTS_LOCK = Mutex.new
-    private_constant :COMPONENTS_LOCK
+    #COMPONENTS_LOCK = Mutex.new
+    #private_constant :COMPONENTS_LOCK

     attr_writer :configuration

@@ -106,7 +106,7 @@ module Datadog
     private

     def safely_synchronize
-      COMPONENTS_LOCK.synchronize do
+      #COMPONENTS_LOCK.synchronize do
         begin
           yield
         rescue ThreadError => e
@@ -117,7 +117,7 @@ module Datadog
           )
           nil
         end
-      end
+      #end
     end
```

**Comment by delner** [comment]
Thanks for doing this performance benchmark! It's great getting some data to get a clearer picture.

> With a single thread, being able to grab the tracer more than 2 million times in the worst case (2.0.0 with mutex) seems reasonable-ish? How many times do we expect a reasonable application to grab the tracer per second?

I think it's a fair point to ask. If we're looking at millions of grabs a second, I think that's a big threshold considering we're unlikely to generate that many spans per second. 

> Kidding aside, there's definitely an impact, but this is quite a big corner case -- 100 threads, each doing nothing but looking and trying to get the tracer.

All-in-all, I'm not as worried about a proportional impact (e.g. 300% slower) as I am practical impact (number of actual retrievals/sec): if those numbers remain well above a common-sense threshold, then I call that a pass, and am content to opt for the simpler, cleaner solution.

My read is that these perf numbers are above that threshold; what do you think?

**Comment by ivoanjo** [comment]
> My read is that these perf numbers are above that threshold; what do you think?

I agree. Even for 2.0.0 with mutex, 100 threads * 2500 acquires * 3.7 = 925000 per second which still seems to be a lot more than we expect.

**Review by marcotc — APPROVED** [review]

**Comment by marcotc** [comment]
I think we should backport this to master when merged.
